### PR TITLE
Added support for weights to mean, median and mode imputer

### PR DESCRIPTION
**Pull request by davidhopkinson26**
Added support for weights to mean, median and mode imputer and updated tests.

Added check_weights_column to BaseTransformer to facilitate validation of weights columns.

### CODE REVIEW

**Review by Ritunjai-Sharma — COMMENTED** [review]
I have made comments regarding the pytest test_weight_not_str_error of mode imputer. Apart from that, I noticed that the tests test_fit_returns_self_weighted, test_fit_not_changing_data_weighted, test_nulls_imputed_correctly_3 and test_learnt_values_not_modified_weights have not been defined for mode imputer but have been defined for mean and median imputers. Wouldn't it be better to define them for mode imputer as well for having consistency across tests for mean, median and mode imputers?

File: tests/imputers/test_ModeImputer.py [hunk 1]
```diff
@@ -51,6 +54,16 @@ def test_super_init_called(self, mocker):
...
+    def test_weight_not_str_error(self):
+        """Test that an exception is raised if weight is not str or None"""
+
+        with pytest.raises(
+            ValueError,
+            match="ModeImputer: weight should be a single column \(str\)",  # noqa
+        ):
+
+            ModeImputer(weight=0)
+
```
[the remark this answers is not in the document]
Comment (davidhopkinson26) [reply tests/imputers/test_ModeImputer.py]: addressed in 9664382 and 727111a

File: tests/imputers/test_ModeImputer.py [hunk 2]
```diff
@@ -51,6 +54,17 @@ def test_super_init_called(self, mocker):
 
             ModeImputer(columns=None, verbose=True, copy=True)
 
+    @pytest.mark.parametrize("weight", (0, ["a"], {"a": 10}))
+    def test_weight_arg_errors(self, weight):
+        """Test that appropriate errors are throw for bad weight arg"""
```
Comment (Ritunjai-Sharma) [inline tests/imputers/test_ModeImputer.py:59]: Super small change: 'throw' should be changed to 'thrown'

**Review by davidhopkinson26 — COMMENTED** [review]
corrected spelling error in docstring

**Review by Ritunjai-Sharma — APPROVED** [review]

**Pull request merged**